rtnetlink/src/ns.rs

```diff
@@ -0,0 +1,301 @@
+use tokio::task;
+
+use crate::Error;
+use nix::{
+    fcntl::OFlag,
+    sched::CloneFlags,
+    sys::{
+        stat::Mode,
+        wait::{waitpid, WaitStatus},
+    },
+    unistd::{fork, ForkResult},
+};
+use std::{option::Option, path::Path, process::exit};
+
+pub const NETNS_PATH: &str = "/run/netns/";
+pub const SELF_NS_PATH: &str = "/proc/self/ns/net";
+pub const NONE_FS: &str = "none";
+
+pub struct NetworkNamespace();
+
+impl NetworkNamespace {
+    /// Add a new network namespace.
+    /// This is equivalent to `ip netns add NS_NAME`.
+    pub async fn add(ns_name: String) -> Result<(), Error> {
+        // Forking process to avoid moving caller into new namespace
+        NetworkNamespace::prep_for_fork()?;
+        log::trace!("Forking...");
+        match unsafe { fork() } {
+            Ok(ForkResult::Parent { child, .. }) => NetworkNamespace::parent_process(child),
+            Ok(ForkResult::Child) => {
+                let netns_path = NetworkNamespace::child_process(ns_name)?;
+                NetworkNamespace::unshare_processing(netns_path)?;
+                exit(0)
+            }
+            Err(e) => {
+                let err_msg = format!("Fork failed: {}", e);
+                Err(Error::NamespaceError(err_msg))
+            }
+        }
+    }
+
+    /// Remove a network namespace
+    /// This is equivalent to `ip netns del NS_NAME`.
+    pub async fn del(ns_name: String) -> Result<(), Error> {
+        let res = task::spawn_blocking(move || {
+            let mut netns_path = String::new();
+            netns_path.push_str(NETNS_PATH);
+            netns_path.push_str(&ns_name);
+            let ns_path = Path::new(&netns_path);
+
+            if nix::mount::umount2(ns_path, nix::mount::MntFlags::MNT_DETACH).is_err() {
+                let err_msg = String::from("Namespace unmount failed (are you running as root?)");
+                return Err(Error::NamespaceError(err_msg));
+            }
+
+            if nix::unistd::unlink(ns_path).is_err() {
+                let err_msg =
+                    String::from("Namespace file remove failed (are you running as root?)");
+                return Err(Error::NamespaceError(err_msg));
+            }
+            Ok(())
+        });
+        match res.await {
+            Ok(r) => r,
+            Err(e) => {
+                let err_msg = format!("Namespace removal failed: {}", e);
+                Err(Error::NamespaceError(err_msg))
+            }
+        }
+    }
+
+    pub fn prep_for_fork() -> Result<(), Error> {
+        // Placeholder function, nothing to do here.
+        Ok(())
+    }
+
+    /// This is the parent process form the fork, it waits for the
+    /// child to exit properly
+    pub fn parent_process(child: nix::unistd::Pid) -> Result<(), Error> {
+        log::trace!("parent_process child PID: {}", child);
+        log::trace!("Waiting for child to finish...");
+        match waitpid(child, None) {
```

Comment:
I just realized that this is actually blocking for as long as the child process is alive. There's a [flag](https://docs.rs/nix/0.19.1/nix/sys/wait/struct.WaitPidFlag.html#associatedconstant.WNOHANG) to make this non-blocking but then we'd have to integrate this with tokio... The more I think about it and the more I think we might just go for the blocking version anyway. Nowadays, `/var/run` is a tmpfs in memory anyway, so `mount` and `mkdir` should be fast enough.

Reply:
Indeed, that is the part I'm really not happy about, but tested it a lot of times on different machines and it was always fast enough to not block all the process.
Indeed it should be integrated with tokio/smol or have this API as blocking.